add rosparam naming conventions [REP 146] (#104)
ros-infrastructure/rep: rep-0146.rst

REP: 146
Title: ROS Robot Parameter Naming Convention
Author: Kei Okada <k-okada@jsk.t.u-tokyo.ac.jp>
Status: Draft
Type: Informational
Content-Type: text/x-rst
Created: 31-Jly-2015
Post-History:

Abstract
========

This REP aims to define the best practices for robot related parameter

> name sapce for making robot-agnostic packages.

Reviewer: sapce -> space.

Motivation
==========

As the number of ROS-supporeted robot increases, we will write
robot-agnostics packages, but still they need robot-specific

> informations as calibration parametesrs.

Reviewer: supporeted -> supported
robot -> robots
agnostics -> agnostic
informations -> information
parametesrs -> parameters

A `robot_description` parameter is widly shared naming convention
among ROS cummnity and this REP proposes such conventions for robot specific information on

> parameter, so application program can build their program on top of that.

Reviewer: widly -> widely
cummnity -> community
parameter -> parameters ("robot specific information on parameter")
program -> programs

> parameter, so application program can build their program on top of that.

Author: sorry....

Specifications
==============

> Paramters

Reviewer: Parameters

---------

robot_description
'''''''''''''''''

`robot_description` stores robot kinematics / volumentric information as

> text format, that can be either URDF or COLLADA file format

Reviewer: volumentric -> volumetric

But perhaps use 'geometry'?

> text format, that can be either URDF or COLLADA file format

Reviewer: And I agree with @mikeferguson: mentioning Collada here seems strange. None of the core tools support that.

> text format, that can be either URDF or COLLADA file format

Author: ok, removed collada, but since robot_model library can read both urdf and collada, so i think most of core tools using that library can support that. (rviz, robot_state_publisher ...)
https://github.com/ros/robot_model/blob/indigo-devel/urdf/src/model.cpp#L142

> text format, that can be either URDF or COLLADA file format

Reviewer: I was not aware of that. It might make sense to keep it in then.

REP: 146
Title: ROS Robot Parameter Naming Convention
Author: Kei Okada <k-okada@jsk.t.u-tokyo.ac.jp>
Status: Draft
Type: Informational
Content-Type: text/x-rst
Created: 31-Jly-2015
Post-History:

Abstract
========

> This REP aims to define the best practices for robot related parameter

Reviewer 2: ```suggestion
This REP aims to define best practices for robot-related ROS parameters
```

> name space for making robot-agnostic packages.

Reviewer 2: ```suggestion
for use in robot-agnostic packages.
```

Motivation
==========

> As the number of ROS-supported robot increases, we will write

Reviewer 2: ```suggestion
As the number of ROS-supported robots increases, we will write
```

robot-agnostic packages, but still they need robot-specific

> information as calibration parameters.

Reviewer 2: ```suggestion
information, such as calibration parameters.
```

Author: thanks! i have commited your fix as e76be8a

> A `robot_description` parameter is widely shared naming convention

Reviewer 2: ```suggestion
The `robot_description` parameter is an example of a widely shared naming convention
```

> among ROS community and this REP proposes such conventions for robot specific information on

Reviewer 2: ```suggestion
among the ROS community. This REP proposes similar conventions for robot-specific information to be stored on the
```

> parameter, so application program can build their program on top of that.

Reviewer 2: ```suggestion
parameter server, helping applications that use these parameters to more easily run on a greater range of platforms.
```

Specifications
==============

Parameters
----------

robot_description
'''''''''''''''''

`robot_description` stores robot kinematics / geometry information using

> URDF format.

Reviewer 2: ```suggestion
the URDF format.
```

robot_type
''''''''''

> The `robot_type` should capture the name of the class of robot in a way that a user would understand the scope of the robot. If it's a commercial product it is recommended to use the production model name. ` For example 'pr2' in PR2 robot case [1], 'pepper' in Pepper robot case and 'fetch' in Fetch robot cases.

Reviewer 2: ```suggestion
The `robot_type` should capture the name of the class of robot in a way that a user would understand the scope of the robot. If it's a commercial product it is recommended to use the production model name. For example 'pr2' in PR2 robot case [1], 'pepper' in Pepper robot case and 'fetch' in Fetch robot cases.
```

robot_name
''''''''''

> `robot_name` stores robot name. It should be set to make sure that this robot instance is unique on the system in much that computers are recommended to have unique hostnames. For the case of PR2 robot, the default robot name is 'robot' [1], but other individuals have named 'james', 'gutsy' etc...

Reviewer 2: ```suggestion
`robot_name` stores the robot's unique name. It should be set to make sure that this robot instance is unique on the system, in the same way that computers are recommended to have unique hostnames. For the case of PR2 robot, the default robot name is 'robot' [1], but other individuals have named 'james', 'gutsy' etc...
```

REP: 146
Title: ROS Robot Parameter Naming Convention
Author: Kei Okada <k-okada@jsk.t.u-tokyo.ac.jp>
Status: Draft
Type: Informational
Content-Type: text/x-rst
Created: 31-Jly-2015
Post-History:

Abstract
========

This REP aims to define the best practices for robot related parameter
name space for making robot-agnostic packages.

Motivation
==========

As the number of ROS-supported robot increases, we will write
robot-agnostic packages, but still they need robot-specific
information as calibration parameters.

A `robot_description` parameter is widely shared naming convention
among ROS community and this REP proposes such conventions for robot specific information on
parameter, so application program can build their program on top of that.

Specifications
==============

Parameters
----------

robot_description
'''''''''''''''''

`robot_description` stores robot kinematics / geometry information using
URDF format.

robot/type
''''''''''

> `robot/type` stores robot type, for example 'pr2' in PR2 robot case [1]

Reviewer 3: It would be good to define the type more clearly. One example is not enough to clearly define what should be here.

Author: add Pepper and Fetch case for `robot_type`
I also added example of `robot_name` if you know other PR2 name, please let me know 
% Unfortunately, we JSK calls them pr2, or pr1012...

> robot/name

Reviewer 3: I think that the namespacing here doesn't really make sense. To be able to introspect/discover this field a longer /more unique name would be better. I'd suggest `robot_name` and `robot_type`.

This also gets rid of the ambiguity of what namespace it applies to. The keys will be in the root of the namespace of the robot. So if you're searching for multiple robots, you don't have to look for a `name` field, and then traverse upward looking for a namespace `robot` and then traverse upward once more to find the `robot_description`

Author: thanks for review, although it will differ from original PR2's name http://github.com/pr2-debs/pr2-core/blob/master/root/etc/ros/robot.yaml, but it much make sense

''''''''''

`robot/name` stores robot name, default pr2 name is 'robot' [1]

> robot/info

Reviewer 3: I don't think that this is currently defined well enough to be useful. I see the name and type as equivalents to `lsb_release` type of information. It would be easy to have a command line tool that just lists robots on the system and displays their name and type. But the info could be printed, but not much else.

If we want to keep this field we'd have to define some level of schema to make it useful.

Author: I added 'robot_info' because I'm afraid users willing to add more robot specific information depending on their use-cases, For example pepper robot can have characteristics of languages they speak or users who own the robot. But I do not want to consider/discuss if we should add `robot_language`, `robot_owner` , because that is very difficult and ambiguous and it is not clear if we should add these information as a  ROS-wide convention.

So 'robot_info'  will store information which each-users willing to store as their robot specific parameters.

Reviewer 3: That's exactly my point. If it's going to be whatever the user wants to store, they have the entire parameter server to store anything they want however they want.  Setting these sort of standards is mostly about restricting what is consider proper usage of a key in the parameter server.

If we were to encourage unstructured use of `robot_info` now. In the future if we wanted to develop a standard schema for `robot_info` we would break every user who had already been using it in their own way, thus we'd actually have to then pick a different key for the new standard approach so as not to collide with the "old" standard which wasn't a standard.

> `robot_type` stores robot type in lower cases, for example 'pr2' in PR2 robot case [1], 'pepper' in Pepper robot case and 'fetch' in Fetch robot cases. This could be regarded as manufacturer's model name.

Reviewer 3: Can you add a justification for lower case? It seems that `PR2` would be fine? I could also see manufactures considering adding some level of versioning in here. For example 

http://library.isr.ist.utl.pt/docs/roswiki/Robots(2f)Kingfisher.html vs https://robots.ros.org/kingfisher/ would likely want to be distinguished as something like `Kingfisher M100` vs `Kingfisher` (though it does appear to be renamed the Heron for disambiguation now)

Reviewer 3: I'd suggest adding something like 
```
The `robot_type` should capture the name of the class of robot in a way that a user would understand the scope of the robot. If it's a commercial product it is recommended to use the production model name. ` Then move onto the examples. 
```

Author: I  just to avoid confusion, for example people prefer the code like  `rospy.get_param("robot/name") == "pr2"`  than   `rospy.get_param("robot/name") == "PR2"`.
or, if wen can use `$(param)` tag in future, some one want to change file name from `PR2_configuration.yaml` to `pr2_configuration.yaml`. So if we decide we only allow lower case, we can get rid of these kind of discussion.
But I found that negative cases, which is "AIBO" (old one) and "aibo" (recent ROS version)..... so  I removed the sentence.

> I could also see manufactures considering adding some level of versioning in here.

That's difficult question. Should we use "AIBO" or "AIBO ERS-110"/"AIBO ERS-21"?
In our cases "HRP-2" or "HRP2JSK" (which is student modified version of original HRP-2 robot)...

But I think important thing is that we can programmably get the name of the current robot and run 'case' function to choose appropriate function.

> `robot_name` stores robot name, default pr2 name is 'robot' [1], which is given individual name to the robot. So robot name of 'pr2' robot type could be 'james', 'gutsy' etc...

Reviewer 3: Might I suggest adding something to the effect of `It should be set to make sure that this robot instance is unique on the system in much that computers are recommended to have unique hostnames.`

Author: thanks updated.

''''''''''

> `robot_info` stores any other characteristics of the robot in YAML dict type, including type, serial number, owner name / email address, timezone, language it uses ...

Reviewer 3: You changed one of the two instances of `robot/info` to `robot_info` but I still think that if we're not going to define any schema there's no point in including it in this standard. That's not to say it can't be there, but until we can desctibe what goes inside it then we shouldn't try to add it here. We can always amend/extend/replace this in the future when it's clear what should be in that field.

Author: ok, removed.
